File: ui/pages/checkout_page.py
from playwright.sync_api import Locator, Page

from pages.base_page import BasePage


class CheckoutPage(BasePage):
    """Checkout steps: customer info, overview, and completion actions."""

    def __init__(self, page: Page, base_url: str) -> None:
        super().__init__(page, base_url)
        self._first_name: Locator = page.locator("#first-name")
        self._last_name: Locator = page.locator("#last-name")
        self._postal_code: Locator = page.locator("#postal-code")
        self._continue: Locator = page.locator("#continue")
        self._cancel: Locator = page.locator("#cancel")
        self._finish: Locator = page.locator("#finish")
        self._complete_header: Locator = page.locator(".complete-header")

    def expect_step_one_loaded(self) -> None:
        self.page.wait_for_url("**/checkout-step-one.html")

    def expect_step_two_loaded(self) -> None:
        self.page.wait_for_url("**/checkout-step-two.html")

    def expect_complete_loaded(self) -> None:
        self.page.wait_for_url("**/checkout-complete.html")

    def fill_customer_info(
        self,
        first_name: str,
        last_name: str,
        postal_code: str,
    ) -> None:
        # Keep field entry in one method so tests read like business flow.
        self._first_name.fill(first_name)
        self._last_name.fill(last_name)
        self._postal_code.fill(postal_code)

    def continue_to_overview(self) -> None:
        self._continue.click()

    def cancel_checkout(self) -> None:
        self._cancel.click()

    def finish_order(self) -> None:
        self._finish.click()

    @property
    def error_message(self) -> Locator:
        return self.page.locator("[data-test='error']")

    @property
    def complete_header(self) -> Locator:
        return self._complete_header

File: java/playwright/src/test/java/com/tnd/playwright/pages/CheckoutPage.java
package com.tnd.playwright.pages;

import com.microsoft.playwright.Page;

/**
 * Page object for SauceDemo checkout steps (Playwright Java).
 *
 * <p>Purpose: model customer-info, overview, and completion actions in one place.
 */
public class CheckoutPage extends BasePage {
    /**
     * Creates checkout page object.
     */
    public CheckoutPage(Page page, String baseUrl) {
        super(page, baseUrl);
    }

    /**
     * Populates checkout identity form fields used in happy-path flow.
     */
    public void fillCustomerInfo(String firstName, String lastName, String postalCode) {
        // Keep checkout identity data entry in one reusable action.
        page.locator("#first-name").fill(firstName);
        page.locator("#last-name").fill(lastName);
        page.locator("#postal-code").fill(postalCode);
    }

    /**
     * Submits checkout step one and waits until overview step is reached.
     */
    public void continueToOverview() {
        page.locator("#continue").click();
        page.waitForURL("**/checkout-step-two.html");
    }

    /**
     * Completes purchase from checkout overview.
     */
    public void finishOrder() {
        page.locator("#finish").click();
        page.waitForURL("**/checkout-complete.html");
    }

    /**
     * Returns visible completion banner text for final assertion.
     */
    public String completeHeaderText() {
        return page.locator(".complete-header").innerText().trim();
    }
}
